### PR TITLE
CA-398341: Populate fingerprints of CA certificates on startup

### PR DESCRIPTION
**Pull request by snwoods**
Also CP-51527: Add --force option to pool-uninstall-ca-certificate.

Addresses the issues raised here by @stormi https://github.com/xapi-project/xen-api/issues/5955

### CODE REVIEW

File: ocaml/xapi-cli-server/cli_frontend.ml
Comment (snwoods) [inline ocaml/xapi-cli-server/cli_frontend.ml:381]: pool-certificate-uninstall is deprecated, so should I be updating it or leaving it as-is? My initial instinct was the latter, but they both use the same underlying function Cli_operations.pool_uninstall_ca_certificate and so the --force option functionality is included in pool-certificate-uninstall regardless.

File: ocaml/xapi-cli-server/cli_frontend.ml
Comment (snwoods) [inline ocaml/xapi-cli-server/cli_frontend.ml:386]: What does this Deprecated flag actually do? It doesn't show when you do xe help pool-certificate-uninstall and it didn't show when I ran the command itself.

File: ocaml/xapi/xapi_pool.ml
Comment (snwoods) [inline ocaml/xapi/xapi_pool.ml:1400]: Trying to understand how the legacy names work. Why is the legacy name "certificate_uninstall" still needed (here and in message_forwarding.ml and datamodel_pool.ml if the deprecated xe command doesn't even use it? It calls pool_install_ca_certificate which calls uninstall_ca_certificate, not certificate_uninstall. What is it providing compatibility with, is this for upgrades from much older xapis?

File: ocaml/xapi/certificates.ml
Comment (gangj) [inline ocaml/xapi/certificates.ml:271]: Do we also need to upgrade for the "host" certificate?

File: ocaml/xapi/certificates.ml
Comment (psafont) [inline ocaml/xapi/certificates.ml:289]: Can you add another catch for any exception? This is out of cautiosness. Because the function is called on startup, it cannot raise any exception.

```ocaml
  | exn ->
      Error (`Msg (Printexc.to_string exn))
```

File: ocaml/idl/datamodel_pool.ml
Comment (psafont) [inline ocaml/idl/datamodel_pool.ml:635]: This call is deprecated, please don't modify it. (we have to change the lifecycle, but the last entry is the deprecation!)

File: ocaml/idl/datamodel_host.ml
Comment (psafont) [inline ocaml/idl/datamodel_host.ml:1090]: Please add an entry to the lifecycle stating it has been modified

File: ocaml/idl/datamodel_pool.ml
Comment (psafont) [inline ocaml/idl/datamodel_pool.ml:658]: Please add an entry to the lifecycle, as in the other call

File: ocaml/xapi-cli-server/cli_frontend.ml
Comment (psafont) [reply ocaml/xapi-cli-server/cli_frontend.ml:381]: Please leave it as-is, same as the api call

File: ocaml/xapi-cli-server/cli_frontend.ml
Comment (psafont) [reply ocaml/xapi-cli-server/cli_frontend.ml:387]: This is probably because there's very little autogenerated code. The cli server should let the client print a message, probably. (will it break the internal regression tests, though?)

File: ocaml/xapi/message_forwarding.ml
Comment (psafont) [inline ocaml/xapi/message_forwarding.ml:3729]: Changing `certificate_uninstall` to pass `~force:false` to `uninstall_ca_certificate` allows the deprecated api function to not be modified in any way.

File: ocaml/xapi/certificates.ml
Comment (gangj) [reply ocaml/xapi/certificates.ml:271]: OK, I see, thank you!
Then can we also cover the "host" certificate upgrade here and then we don't need the same functionality in two places.

Comment (psafont) [reply ocaml/xapi/certificates.ml:271]: I don't think it makes sense as it is, currently. It's used for host certificates, exclusively, when the before the stunnel component gets started.

These certificates gets used by the client part, so the update needs to happen at a time that's independent (or risk introducing races) Maybe we can consolidate some of the modules, because the certificate handling is a bit disperse now.

File: ocaml/xapi/certificates.ml
Comment (psafont) [inline ocaml/xapi/certificates.ml:348]:
```suggestion
let host_uninstall kind ~name ~force =
```
There's no need to make the parameter optional. Optional parameters complicate usage in some occasions, because of their interaction with currying; and they are really only useful in situations where compatibility needs to be maintained, or other contrived situations.

In cases where we control all the users, such as this one, it's better to use named arguments

File: ocaml/xapi/certificates.ml
Comment (psafont) [inline ocaml/xapi/certificates.ml:459]:
```suggestion
let pool_uninstall kind ~__context ~name ~force =
```

File: ocaml/xapi/xapi_pool.ml
Comment (snwoods) [inline ocaml/xapi/xapi_pool.ml:1405]: I've swapped the order of these definitions around to match message_forwarding.ml, which also allows us to pass ~force:false so that certificate_uninstall's is unchanged, does this make sense @psafont or was there a reason for them being this way around in xapi_pool.ml?

File: ocaml/idl/datamodel_host.ml
Comment (snwoods) [inline ocaml/idl/datamodel_host.ml:1094]: @psafont Is this the correct thing to do? This is currently the correct version but it could change before release. Does this just need to be updated again before release or is there some way to auto-fill it? This only happens when you have lifecycle:[] right?

File: ocaml/xapi/xapi_pool.ml
Comment (psafont) [reply ocaml/xapi/xapi_pool.ml:1405]: the order of the bindings doesn't matter, whichever is more convenient for you, is good

File: ocaml/xapi/certificates.ml
Comment (snwoods) [reply ocaml/xapi/certificates.ml:271]: So, leave it as it is in this PR for now but with the potential to consolidate the certificate handling code in the future?

File: ocaml/xapi-cli-server/cli_frontend.ml
Comment (snwoods) [reply ocaml/xapi-cli-server/cli_frontend.ml]: I've removed force from both this and `certificate_uninstall` in datamodel_pool.ml, but when testing pool-certificate-uninstall, force still works. This is expected because as I say, `certificate-uninstall` is using `Cli_operations.pool_uninstall_ca_certificate`, the same as `pool-uninstall-ca-certificate`. Is this what you intended? It adds functionality to the deprecated function but doesn't change the documentation.

(datamodel_pool.certificate_uninstall is completed unused AFAICT)

File: ocaml/xapi/certificates.ml
Comment (psafont) [reply ocaml/xapi/certificates.ml:271]: That's right

File: ocaml/xapi/xapi_pool.ml
Comment (psafont) [reply ocaml/xapi/xapi_pool.ml:1400]: API clients might still use this function, this is why it needs to be kept.

**Review by psafont — APPROVED** [review]

File: ocaml/xapi/xapi.ml
Comment (robhoes) [inline ocaml/xapi/xapi.ml:1152]: Is this a one-off upgrade case? That is, once you have the xapi version with this change in it and it has run once, you'll never need it anymore? Then it would better fit in xapi_db_upgrade.ml (bump the schema version and add a < rule).

File: ocaml/idl/datamodel_host.ml
Comment (robhoes) [inline ocaml/idl/datamodel_host.ml:1084]: I think this would break existing clients, because the new param does not have a default. Use `~versioned_params` for that.

**Review by robhoes — CHANGES_REQUESTED** [review]
See above.

**Comment by lindig** [comment]
Can we pick this up again? @snwoods

File: ocaml/idl/datamodel_host.ml
Comment (robhoes) [inline ocaml/idl/datamodel_host.ml:1502]: Please make it `24.32.0`, as we have 31 already released.

File: ocaml/xapi/xapi_db_upgrade.ml
Comment (robhoes) [inline ocaml/xapi/xapi_db_upgrade.ml:910]: Why 779? You have used 783 in the datamodel.

File: ocaml/xapi/xapi_db_upgrade.ml
Comment (robhoes) [inline ocaml/xapi/xapi_db_upgrade.ml:907]: This is dead code until you add the new function to the `rules` list below.

File: ocaml/xapi/xapi_db_upgrade.ml
Comment (snwoods) [reply ocaml/xapi/xapi_db_upgrade.ml:910]: Ah 779 was when the fingerprint_sha256 and fingerprint_sha1 fields were added, so I was thinking anything from that version should already have those fields populated, but of course if they'd previously upgraded from <779 they wouldn't have it, sorry.

**Review by robhoes — APPROVED** [review]

**Review by psafont — CHANGES_REQUESTED** [review]
This contains an unrelated commit from https://github.com/xapi-project/xen-api/pull/6045/

let's wait until that PR is merged

**Comment by snwoods** [comment]
Tested on a toolstack-next host and it worked as expected (updated the fingerprints when updating to this code, but didn't run the upgrade code from then on)

**Comment by psafont** [comment]
Please remove my commit and rebase on top of master, then it should be good to go

**Review by psafont — APPROVED** [review]

**Comment by psafont** [comment]
Let's wait until the metrics fix is in to merge this

**Comment by lindig** [comment]
I believe @psafont the PR you referenced was merged; so we can merge this now?

**Comment by psafont** [comment]
The versions in the datamodel need to be updated to 24.34.0, otherwise it's good to go

**Pull request merged**
